src/diffusers/pipelines/blip_diffusion/pipeline_blip_diffusion.py

```diff
@@ -0,0 +1,337 @@
+
+import numpy as np
+import PIL
+from ...models import AutoencoderKL, UNet2DConditionModel, ControlNetModel
+from .modeling_ctx_clip import CtxCLIPTextModel
+from transformers import CLIPTokenizer
+from ...pipelines import DiffusionPipeline
+import torch
+from ...schedulers import PNDMScheduler
+from ...utils import (
+    BaseOutput,
+    is_accelerate_available,
+    is_accelerate_version,
+    logging,
+    randn_tensor,
+    replace_example_docstring,
+)
+from ...utils.pil_utils import PIL_INTERPOLATION
+from torch import nn
+from transformers.activations import QuickGELUActivation as QuickGELU
+from .modeling_blip2 import Blip2QFormerModel
+import tqdm
+from torchvision import transforms
+from torchvision.transforms.functional import InterpolationMode
+from ...image_processor import BlipImageProcessor
+from PIL import Image
+logger = logging.get_logger(__name__)  # pylint: disable=invalid-name
+import re
+
+def prepare_cond_image(
+        image, width, height, batch_size, device, do_classifier_free_guidance=True
+    ):
+        if not isinstance(image, torch.Tensor):
+            if isinstance(image, Image.Image):
+                image = [image]
+
+            if isinstance(image[0], Image.Image):
+                images = []
+
+                for image_ in image:
+                    image_ = image_.convert("RGB")
+                    image_ = image_.resize(
+                        (width, height), resample=PIL_INTERPOLATION["lanczos"]
+                    )
+                    image_ = np.array(image_)
+                    image_ = image_[None, :]
+                    images.append(image_)
+
+                image = images
+
+                image = np.concatenate(image, axis=0)
+                image = np.array(image).astype(np.float32) / 255.0
+                image = image.transpose(0, 3, 1, 2)
+                image = torch.from_numpy(image)
+            elif isinstance(image[0], torch.Tensor):
+                image = torch.cat(image, dim=0)
+
+        image_batch_size = image.shape[0]
+
+        if image_batch_size == 1:
+            repeat_by = batch_size
+        else:
+            # image batch size is the same as prompt batch size
+            # repeat_by = num_images_per_prompt
+            raise NotImplementedError
+
+        image = image.repeat_interleave(repeat_by, dim=0)
+
+        # image = image.to(device=self.device, dtype=dtype)
+        image = image.to(device=device)
+
+        if do_classifier_free_guidance:
+            image = torch.cat([image] * 2)
+
+        return image
+
+
+
+# Create a class for the Blip Diffusion pipeline
+class BlipDiffusionPipeline(DiffusionPipeline):
+    
+    def __init__(self, tokenizer: CLIPTokenizer, text_encoder: CtxCLIPTextModel, vae: AutoencoderKL, unet: UNet2DConditionModel, scheduler: PNDMScheduler, qformer: Blip2QFormerModel, controlnet: ControlNetModel=None, ctx_begin_pos: int = 2):
+        super().__init__()
+
+        self.image_processor = BlipImageProcessor()
```

Comment:
The `image_processor` also needs to be registered.

Reply:
Does `ControlNet` always need to be present to use this pipeline? I think it'd be better to create two pipelines here:

* `BlipDiffusionPipeline`
* BlipDiffusionControlNetPipeline`

Reply:
Done, created a new `BlipDiffusionControlNetPipeline`